File: master/client/assets/script/game/entity/MineAction.ts
import Item from "./Item";

export type MineAction =
    | MineActionMinerChangeDirection
    | MineActionMinerMove
    | MineActionMinerHitBrick
    | MineActionBrickDamage
    | MineActionShowBrickReward
    | MineActionShowDiamondWindow
    | MineActionDelay
    | MineActionFloorChange
    | MineActionShowBoss
    | MineActionBattle;

type MineActionMinerChangeDirection = {
    action: "minerChangeDirection";
    /** 面向左或者面向右 */
    data: { direction: "left" | "right" };
};

type MineActionMinerMove = {
    action: "minerMove";
    /** 矿工移动 */
    data: { x: number; y: number };
};

type MineActionMinerHitBrick = {
    action: "minerHitBrick";
    /** 矿工击打砖块 */
    data: { uniqueId: number };
};

type MineActionBrickDamage = {
    action: "brickDamage";
    /** 砖块受到伤害，life指剩余生命值 */
    data: { uniqueId: number; life: number };
};

type MineActionShowBrickReward = {
    action: "showBrickReward";
    /** 显示砖块奖励 */
    data: { uniqueId: number; reward: Item[] };
};

type MineActionShowDiamondWindow = {
    action: "showDiamondWindow";
    /** 显示钻石窗口 */
    data: { count: number };
};

type MineActionDelay = {
    action: "delay";
    /** 延迟，单位毫秒 */
    data: { time: number };
};

type MineActionFloorChange = {
    action: "floorChange";
    /** 楼层变化，整个地图需要向上滚 */
    data: { floor: number };
};

type MineActionShowBoss = {
    action: "showBoss";
    /** 显示boss */
    data: { bossId: number };
};

type MineActionBattle = {
    action: "battle";
    /** 战斗 */
    data: { uniqueId: number };
};

File: master/server_main/server_game_module/src/Game/Entity/MineAction.cs
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
using System.Threading.Tasks;

namespace GamePlay
{
    public record MineAction(string action, object data) { }

    public record MineActionMinerChangeDirection(string direction)
    {
        public MineAction ToAction()
        {
            return new MineAction("minerChangeDirection", this);
        }
    }

    public record MineActionMinerMove(int x, int y)
    {
        public MineAction ToAction()
        {
            return new MineAction("minerMove", this);
        }
    }

    public record MineActionMinerHitBrick(int uniqueId)
    {
        public MineAction ToAction()
        {
            return new MineAction("minerHitBrick", this);
        }
    }

    public record MineActionBrickDamage(int uniqueId, int life)
    {
        public MineAction ToAction()
        {
            return new MineAction("brickDamage", this);
        }
    }

    public record MineActionShowBrickReward(int uniqueId, ImmutableArray<Item> reward)
    {
        public MineAction ToAction()
        {
            return new MineAction("showBrickReward", this);
        }
    }

    public record MineActionShowDiamondWindow(int count)
    {
        public MineAction ToAction()
        {
            return new MineAction("showDiamondWindow", this);
        }
    }

    public record MineActionDelay(long time)
    {
        public MineAction ToAction()
        {
            return new MineAction("delay", this);
        }
    }

    public record MineActionFloorChange(int floor)
    {
        public MineAction ToAction()
        {
            return new MineAction("floorChange", this);
        }
    }

    public record MineActionShowBoss(int bossId)
    {
        public MineAction ToAction()
        {
            return new MineAction("showBoss", this);
        }
    }

    public record MineActionBattle(int uniqueId)
    {
        public MineAction ToAction()
        {
            return new MineAction("battle", this);
        }
    }
}